// @ained/composables/index.ts
import { useCase } from './useCase.js'
import { useClipboard } from './useClipboard.js'
import { useColor } from './useColor.js'
import { useDebouncer } from './useDebounce.js'
import { useDragger } from './useDragger.js'
import { useBadge } from './useBadge.js'
import { useMediaQuery } from './useMediaQuery.js'
import { useFuzzy } from './useFuzzy.js'
import { useUUID } from './useUUID.js'
import { useCurrency } from './useCurrency.js'
import { usePureClick } from './usePureClick.js'
import { useShare } from './useShare.js'
import { useDeeplink } from './useDeeplink.js'
import { useTweened } from './useTweened.js'

import { clone } from './utils/clone.js'
import { between } from './utils/between.js'
import { polyfills } from './utils/polyfills.js'
import { regexp } from './utils/regexp.js'
import { empty } from './utils/empty.js'
import { filter } from './utils/filter.js'
import { prototype } from './utils/prototype.js'
import { easing } from './utils/easing.js'

export {
    useCase,
    useClipboard,
    useColor,
    useDebouncer,
    useDragger,
    useBadge,
    useMediaQuery,
    useFuzzy,
    useUUID,
    useCurrency,
    usePureClick,
    useShare,
    useDeeplink,
    useTweened,

    clone, between, polyfills, regexp, empty, filter, prototype, easing
}
